stl/src/asan.cpp

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cstdint>
+
+extern "C" {
+void __sanitizer_annotate_contiguous_container(
+    const void* _First, const void* _End, const void* _Old_last, const void* _New_last) noexcept;
+
+void _Asan_vector_aligned_annotate_start(const void* _Vector_first, const void* _Vector_last, const void* _Vector_end,
+    const size_t _Vector_new_size, const size_t _Type_size, const void** _Asan_last_out) {
+    if (!_Vector_first) {
+        // The vector is empty. We'll extend the accessible range in the destructor.
+        return;
+    }
+
+    const char* _First     = reinterpret_cast<const char*>(_Vector_first);
+    const char* _Last      = reinterpret_cast<const char*>(_Vector_last);
+    const char* _End       = reinterpret_cast<const char*>(_Vector_end);
+    const size_t _New_size = _Vector_new_size * _Type_size;
+
+    if (_New_size > static_cast<size_t>(_End - _First)) {
+        // New size is larger than current capacity; assume we're allocating a new block.
+        // Clean up the shadow memory annotations for the buffer being freed.
+        __sanitizer_annotate_contiguous_container(_First, _End, _Last, _End);

```

Comment:
I would rather do all the work in the destructor and only cache the `_Old_first`, `_Old_last` and `_Old_end`

That way we would know whether we did reallocate or how many new elements there are

Reply:
Sure, you're right we can do this cleanup in the destructor. I did it here because I was thinking that the pointers could become invalid later but we're not actually ever de-referencing them, we just use them as a key into the shadow memory.

Reply:
Also doing this cleanup assuming the happy case of no throwing (so that the reallocation actually does occur) allows us to cache fewer variables. I really just need to capture the current end of the allowed region of access as far as asan is concerned.

Reply:
I just tried this change and for the simpler functions like `push_back` I see perf regressions even worse than the current implementation.

Reply:
Could you provide the test program? I am a bit swamped right now due to school starting over here but would want to play around a bit soonish